File: templates/dotnet/PersistentSubscriptionExample.cs
// KurrentDB .NET Client Example - Persistent Subscriptions
// This file is a reference example - copy and adapt for your use case

using System.Text;
using System.Text.Json;
using KurrentDB.Client;

namespace KurrentDbExample;

/// <summary>
/// Demonstrates persistent subscriptions with KurrentDB.
/// Run from Program.cs by calling: await PersistentSubscriptionDemo.RunAsync();
/// </summary>
public static class PersistentSubscriptionDemo
{
    public static async Task RunAsync()
    {
        // === CONNECTION ===
        var connectionString = Environment.GetEnvironmentVariable("KURRENTDB_CONNECTION_STRING")
            ?? "kurrentdb://localhost:2113?tls=false";

        var settings = KurrentDBClientSettings.Create(connectionString);
        await using var client = new KurrentDBClient(settings);
        await using var psClient = new KurrentDBPersistentSubscriptionsClient(settings);

        Console.WriteLine($"Connected to KurrentDB at {connectionString}");

        var streamName = "orders";
        var groupName = "order-processor";

        // === CREATE PERSISTENT SUBSCRIPTION ===
        try
        {
            await psClient.CreateToStreamAsync(
                streamName,
                groupName,
                new PersistentSubscriptionSettings(),
                userCredentials: new UserCredentials("admin", "changeit")
            );
            Console.WriteLine($"Created persistent subscription '{groupName}' on stream '{streamName}'");
        }
        catch (Exception ex) when (ex.Message.Contains("already exists"))
        {
            Console.WriteLine($"Persistent subscription '{groupName}' already exists");
        }

        // === APPEND SOME TEST EVENTS ===
        for (int i = 0; i < 3; i++)
        {
            var orderCreated = new { OrderId = Guid.NewGuid().ToString(), Amount = 10.00m * (i + 1) };
            var eventData = new EventData(
                Uuid.NewUuid(),
                "OrderCreated",
                JsonSerializer.SerializeToUtf8Bytes(orderCreated)
            );
            await client.AppendToStreamAsync(streamName, StreamState.Any, new[] { eventData });
            Console.WriteLine($"Appended OrderCreated event #{i + 1}");
        }

        // === SUBSCRIBE TO PERSISTENT SUBSCRIPTION ===
        Console.WriteLine("\nSubscribing to persistent subscription...");

        using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(30));
        var processedCount = 0;

        try
        {
            await using var subscription = psClient.SubscribeToStream(
                streamName,
                groupName,
                cancellationToken: cts.Token
            );

            await foreach (var message in subscription.Messages.WithCancellation(cts.Token))
            {
                switch (message)
                {
                    case PersistentSubscriptionMessage.Event(var evnt, var retryCount):
                        try
                        {
                            Console.WriteLine($"  Processing: {evnt.Event.EventType} (retry: {retryCount})");
                            Console.WriteLine($"  Data: {Encoding.UTF8.GetString(evnt.Event.Data.Span)}");

                            // Simulate processing
                            await Task.Delay(100);

                            // Acknowledge success
                            await subscription.Ack(evnt);
                            Console.WriteLine($"  Acknowledged event");

                            processedCount++;
                            if (processedCount >= 3)
                            {
                                Console.WriteLine("\nAll tests passed!");
                                return;
                            }
                        }
                        catch (Exception ex)
                        {
                            // Park failed event
                            await subscription.Nack(
                                PersistentSubscriptionNakEventAction.Park,
                                ex.Message,
                                evnt
                            );
                            Console.WriteLine($"  Parked event: {ex.Message}");
                        }
                        break;

                    case PersistentSubscriptionMessage.SubscriptionConfirmation confirmation:
                        Console.WriteLine($"  Subscription confirmed: {confirmation.SubscriptionId}");
                        break;
                }
            }
        }
        catch (OperationCanceledException)
        {
            Console.WriteLine("\nSubscription timed out.");
        }
    }
}

File: templates/fsharp/PersistentSubscriptionExample.fs
// KurrentDB F# Client Example - Persistent Subscriptions

open System
open System.Text
open System.Text.Json
open System.Threading
open KurrentDB.Client

// === EVENT TYPES ===
type OrderCreated = {
    OrderId: string
    Amount: decimal
}

let main () =
    task {
        // === CONNECTION ===
        let connectionString =
            Environment.GetEnvironmentVariable("KURRENTDB_CONNECTION_STRING")
            |> Option.ofObj
            |> Option.defaultValue "kurrentdb://localhost:2113?tls=false"

        let settings = KurrentDBClientSettings.Create(connectionString)
        use client = new KurrentDBClient(settings)
        use psClient = new KurrentDBPersistentSubscriptionsClient(settings)

        printfn "Connected to KurrentDB at %s" connectionString

        let streamName = "orders"
        let groupName = "order-processor"

        // === CREATE PERSISTENT SUBSCRIPTION ===
        try
            do! psClient.CreateToStreamAsync(
                streamName,
                groupName,
                PersistentSubscriptionSettings(),
                userCredentials = UserCredentials("admin", "changeit")
            )
            printfn "Created persistent subscription '%s' on stream '%s'" groupName streamName
        with ex when ex.Message.Contains("already exists") ->
            printfn "Persistent subscription '%s' already exists" groupName

        // === APPEND SOME TEST EVENTS ===
        for i in 0..2 do
            let order = {
                OrderId = Guid.NewGuid().ToString()
                Amount = decimal (10.0 * float (i + 1))
            }
            let eventData =
                EventData(
                    Uuid.NewUuid(),
                    "OrderCreated",
                    JsonSerializer.SerializeToUtf8Bytes(order)
                )
            let! _ = client.AppendToStreamAsync(streamName, StreamState.Any, [| eventData |])
            printfn "Appended OrderCreated event #%d" (i + 1)

        // === SUBSCRIBE TO PERSISTENT SUBSCRIPTION ===
        printfn "\nSubscribing to persistent subscription..."

        use cts = new CancellationTokenSource(TimeSpan.FromSeconds(30.0))
        let mutable count = 0
        let mutable continueLoop = true

        try
            use subscription = psClient.SubscribeToStream(streamName, groupName, cancellationToken = cts.Token)

            while continueLoop do
                let! moveNext = subscription.Messages.MoveNextAsync(cts.Token)
                if moveNext then
                    match subscription.Messages.Current with
                    | :? PersistentSubscriptionMessage.Event as evtMsg ->
                        let evnt = evtMsg.ResolvedEvent
                        let retryCount = evtMsg.RetryCount

                        try
                            printfn "  Processing: %s (retry: %d)" evnt.Event.EventType retryCount
                            printfn "  Data: %s" (Encoding.UTF8.GetString(evnt.Event.Data.Span))

                            // Simulate processing
                            // ... your business logic here ...

                            // Acknowledge success
                            do! subscription.Ack(evnt)
                            printfn "  Acknowledged event"
                        with ex ->
                            // Park failed event
                            do! subscription.Nack(PersistentSubscriptionNakEventAction.Park, ex.Message, evnt)
                            printfn "  Parked event: %s" ex.Message

                        count <- count + 1
                        if count >= 3 then
                            continueLoop <- false
                    | :? PersistentSubscriptionMessage.SubscriptionConfirmation as confirm ->
                        printfn "  Subscription confirmed: %s" confirm.SubscriptionId
                    | _ -> ()
                else
                    continueLoop <- false
        with
        | :? OperationCanceledException ->
            printfn "Subscription timed out."

        printfn "\nAll tests passed!"
    }
    |> Async.AwaitTask
    |> Async.RunSynchronously
